### ISSUE TITLE
Evaluating MMIO operations in OPTEE OS vs Linux kernel modules.

### ISSUE THREAD

**Issue by Yuhala** [OP-TEE/optee_os#6235]
Hi all,

I am writing an OPTEE OS kernel driver which performs some MMIO operations with routines like: `readl`, `writel`, etc. Recently, I  have done some benchmarking of these operations in my OPTEE OS kernel driver and compared to the same operations in a normal Linux kernel module (loaded with `insmod`). My preliminary evaluations have interesting results which I'm not sure how to explain. For example, `readl` in the Linux kernel module takes about `37 CPU cycles` (give or take). In my OPTEE OS kernel driver `readl` takes about `27 CPU cycles` for the same MMIO register. Also, `writel` in the Linux module takes about `23 CPU cycles`. I am yet to benchmark `writel` in the OPTEE OS kernel driver (having a few bugs/kernel panics). 

I would have expected `readl` in the Linux kernel module and that in OPTEE OS kernel driver to have similar overall duration. Similarly, I would have expected `writel` to be a bit slower with respect to `readl`. I would be grateful if anyone could give a few reasons why I get these results and if they are normal. 

I am doing my tests on an Nvidia Jetson AGX Xavier system (ARMv8.2 64 bit) which runs Jetson Linux 35.3.1 (based on Linux Kernel 5.10). This package comes with OPTEE and I'm not sure of the exact OPTEE OS commit number(I will update once I get it). The number of cycles for the operations is measured by reading the `cntvct_el0` register before and after the operation and calculating the difference. All operations are done multiple times and the average value calculated.

Thanks.

**Comment by jenswi-linaro** [comment]
That's curious. Are you using the same instructions to implement `readl()` and `writel()`?
You say average. I guess that means all addresses should be cached in the TLB. Or perhaps that's the difference, for some reason not all addresses needed can be cached at the same time in the Linux kernel module?
Do you use a hypervisor in the normal world, that is, is non-secure EL2 enabled?

**Comment by Yuhala** [comment]
Hi @jenswi-linaro , thanks for your response. I did some digging regarding the actual instructions called and noticed some slight differences in the `readl`/`write` implementations for Linux and OPTEE OS.  After reimplementing my benchmarks to use the same implementations for both systems, both systems now have similar `readl` latencies. Also I noticed a bug in my benchmarking calculations which explains the unusual high cycle count value for a `readl` or `writel` instruction. I am closing this issue.

Thanks for your time.
